httm.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH HTTM "1" "June 2026" "httm 0.50.2" "User Commands"
.SH NAME
httm \- manual page for httm 0.50.2
.SH SYNOPSIS
.B httm
[\fI\,OPTIONS\/\fR] [\fI\,INPUT_FILES\/\fR]...
.SH DESCRIPTION
httm prints the size, date and corresponding locations of available unique versions of files residing on snapshots. May also be used interactively to select and restore from such versions, and even to snapshot datasets which contain certain files.
.SS "Arguments:"
.TP
[INPUT_FILES]...
in the non\-interactive modes (when BROWSE, SELECT, or RESTORE are not specified), user may specify one or many paths for a simple display of snapshot versions.  If no paths are included as arguments, then httm will pause waiting for paths to be piped in via stdin (e.g. 'find . | httm').  In the interactive modes (when BROWSE, SELECT, or RESTORE are specified), user may specify one base directory from which to begin a search.  If the interactive modes, if no directory is specified, httm will use the current working directory.
.SH OPTIONS
.TP
\fB\-b\fR, \fB\-\-browse\fR
interactive browse and search a specified directory to display unique file versions. [aliases: \fB\-\-interactive]\fR
.TP
\fB\-s\fR, \fB\-\-select[=\fR<SELECT>]
interactive browse and search a specified directory to display unique file versions. Continue to another dialog to select a snapshot version to dump to stdout. This argument optionally takes a value. Default behavior/value is to simply print the path name, but, if the path is a file, the user can print the file's contents by giving the value "contents", or print any PREVIEW output by giving the value "preview". [possible values: path, contents, preview]
.TP
\fB\-c\fR, \fB\-\-copy[=\fR<COPY>]
interactive browse and search a specified directory to display unique file versions. Continue to another dialog to select a snapshot version to copy. This argument optionally takes a value. Default behavior/value is a non\-destructive "copy" to the current working directory with a new name, so as not to overwrite any "live" file version. However, the user may specify the "preserve" value.  Preserve mode will attempt to preserve attributes, like the permissions/mode, timestamps, xattrs and ownership of the selected snapshot file version.  User may also set the copy/restore mode via the HTTM_RESTORE_MODE environment variable. [possible values: copy, copy\-and\-preserve, preserve]
.TP
\fB\-r\fR, \fB\-\-restore[=\fR<RESTORE>]
interactive browse and search a specified directory to display unique file versions. Continue to another dialog to select a snapshot version to restore. This argument optionally takes a value. Default behavior/value is "overwrite" (or "yolo") to restore to the same file location. Note, "overwrite" can be a DESTRUCTIVE operation. Overwrite mode will attempt to preserve attributes, like the permissions/mode, timestamps, xattrs and ownership of the selected snapshot file version. User may also specify "guard".  Guard mode has the same semantics as "overwrite" but will attempt to take a precautionary snapshot before any overwrite action occurs. Note: Guard mode is a ZFS only option. User may also set the copy/restore mode via the HTTM_RESTORE_MODE environment variable. [possible values: overwrite, yolo, guard]
.TP
\fB\-d\fR, \fB\-\-deleted[=\fR<DELETED>]
show deleted files in interactive modes (BROWSE, SELECT and RESTORE). When an interactive mode is not specified, search for all files deleted from a specified directory. This argument optionally takes a value. When specified, the default behavior/value is "all". If "only" is specified, then, in the interactive modes, non\-deleted files will be excluded from the search. If "single" is specified, then, deleted files behind deleted directories, (files with a depth greater than one) will be ignored. [possible values: all, single, only, one]
.TP
\fB\-R\fR, \fB\-\-recursive\fR
recurse into the selected directory to find more files and directories. Only available in interactive (BROWSE, SELECT and RESTORE) and DELETED file modes.
.TP
\fB\-a\fR, \fB\-\-alt\-replicated\fR
automatically discover locally replicated datasets and list their snapshots as well. NOTE: Be certain such replicated datasets are mounted before use. httm will silently ignore unmounted datasets in the interactive modes.
.TP
\fB\-p\fR, \fB\-\-preview[=\fR<PREVIEW>]
user may specify a command to preview snapshots while in a snapshot selection view.  This argument optionally takes a value specifying the command to be executed.  The default value/command, if no command value specified, is a 'bowie' formatted 'diff'.  User defined commands must specify the snapshot file name "{snap_file}" and the live file name "{live_file}" within their shell command. NOTE: 'bash' is required to bootstrap any preview script, even if the user specifies their own preview command, written in a different shell language.
.TP
\fB\-\-dedup\-by[=\fR<DEDUP_BY>]
comparing file versions solely on the basis of size and modify time (the default "metadata" behavior) may return what appear to be "false positives and negatives".  This is because metadata, specifically modify time and size, is not a precise measure of whether a file has actually changed. A program might overwrite a file with the same contents, and/or a user can simply update the modify time via 'touch'. This flag, when specified with the "contents" option, compares the actual file contents of same\-sized file versions, overriding the default "metadata" only behavior. The "contents" option is file\-only (it does not apply to non\-file objects, like directories) and can be expensive, as the file versions need to be read back and compared.  Given how time expensive this operation can be if specified, the "contents" option is not shown in BROWSE mode, but after a selection is made, can be utilized, when enabled, in SELECT or RESTORE modes.  A less expensive, "suspect" option, only compares file contents when a file's metadata makes it likely the file may have new file contents, such as when that file has a new inode or birth time.  The "disable" "all" or "no\-filter" option dumps all snapshot versions, without deduplication. [aliases: \fB\-\-unique\fR, \fB\-\-uniqueness]\fR [possible values: disable, all, no\-filter, metadata, contents, suspect]
.TP
\fB\-e\fR, \fB\-\-exact\fR
use exact pattern matching for searches in the interactive modes (in contrast to the default fuzzy searching).
.TP
\fB\-S\fR, \fB\-\-snap[=\fR<SNAPSHOT>]
snapshot a file/s most immediate mount. This argument optionally takes a value for a snapshot suffix. The default suffix is 'httmSnapFileMount'. Note: This is a ZFS only option which requires either superuser or 'zfs allow' privileges.
.TP
\fB\-\-list\-snaps[=\fR<LIST_SNAPS>]
display snapshots names for a file. This argument optionally takes a value. By default, this argument will return all available snapshot names. When the DEDUP_BY flag is not specified, but LIST_SNAPS is, the default DEDUP_BY level is "all" snapshots. Thus the user may limit type of snapshots returned via specifying the DEDUP_BY flag. The user may also omit the most recent "n" snapshots from any list. By appending a comma, this argument also filters those snapshots which contain the specified pattern/s. A value of "5,prep_Apt" would return the snapshot names of only the last 5 (at most) of all snapshot versions which contain "prep_Apt". The value "native" will restrict selection to only 'httm' native snapshot suffix values, like "httmSnapFileMount" and "ounceSnapFileMount". Note: This is a ZFS and btrfs only option.
.TP
\fB\-\-roll\-forward=\fR<ROLL_FORWARD>
whereas 'zfs rollback' is a destructive operation, this operation is non\-destructive. This operation preserves interstitial snapshots, and requires a snapshot guard before taking any action.  If this flag is specified (along with the required snapshot name), httm will modify (copy and delete) those files and their attributes (preserving hard links) that have changed since the specified snapshot to the live dataset. httm will also take two precautionary guard snapshots, one before and one after the operation. Should the roll forward fail for any reason, httm will rollback to the pre\-execution state. CAVEATS: This is a ZFS only option which requires super user privileges.  Not all filesystem features are supported (for instance, Unix sockets on the snapshot will need to be recreated) and may cause a roll forward to fail.  Moreover, certain special objects/files will be copied or recreated, but are not guaranteed to be in the same state as the snapshot (for instance, FIFO buffers).
.TP
\fB\-\-prune\fR
prune all snapshot/s which contain the input file/s on that file's most immediate mount via "zfs destroy". "zfs destroy" is a DESTRUCTIVE operation which DOES NOT ONLY APPLY to the file in question, but the entire snapshot upon which it resides. Careless use may cause you to lose snapshot data you care about. This argument requires and will be filtered according to any values specified at LIST_SNAPS. User may also enable SELECT mode to make a more granular selection of specific snapshots to prune. Note: This is a ZFS only option.
.TP
\fB\-m\fR, \fB\-\-file\-mount[=\fR<FILE_MOUNT>]
by default, display the all mount point/s of all dataset/s which contain/s the input file/s. This argument optionally takes a value to display other information about the path. Possible values are: "mount" or "target" or "directory", the default value, returns the mount/directory of a file's underlying dataset, "source" or "device" or "dataset", returns a file's underlying dataset/device, and, "relative\-path" or "relative", returns a file's relative path from the underlying mount. [aliases: \fB\-\-mount]\fR [possible values: source, target, mount, directory, device, dataset, relative\-path, relative, relpath]
.TP
\fB\-l\fR, \fB\-\-last\-snap[=\fR<LAST_SNAP>]
automatically select and print the path of last\-in\-time unique snapshot version for the input file. This argument optionally takes a value. Possible values are: "any", return the last in time snapshot version, this is the default behavior/value, "ditto", return only last snaps which are the same as the live file version, "no\-ditto\-exclusive", return only a last snap which is not the same as the live version (argument "\-\-no\-ditto" is an alias for this option), "no\-ditto\-inclusive", return a last snap which is not the same as the live version, or should none exist, return the live file, and, "none" or "without", return the live file only for those files without a last snapshot. [aliases: \fB\-\-last\fR, \fB\-\-latest]\fR [possible values: any, ditto, no\-ditto, no\-ditto\-exclusive, no\-ditto\-inclusive, none, without]
.TP
\fB\-n\fR, \fB\-\-raw\fR
display the snapshot locations only, without extraneous information, delimited by a NEWLINE character. [aliases: \fB\-\-newline]\fR
.TP
\fB\-0\fR, \fB\-\-zero\fR
display the snapshot locations only, without extraneous information, delimited by a NULL character. [aliases: \fB\-\-null]\fR
.TP
\fB\-\-csv\fR
display all information, delimited by a comma.
.TP
\fB\-\-not\-so\-pretty\fR
display the ordinary output, but tab delimited, without any pretty border lines. [aliases: \fB\-\-tabs\fR, \fB\-\-plain\-jane\fR, \fB\-\-not\-pretty]\fR
.TP
\fB\-\-json\fR
display the ordinary output, but as formatted JSON.
.TP
\fB\-\-omit\-ditto\fR
omit display of the snapshot version which may be identical to any live version. By default, `httm` displays all snapshot versions and the live version).
.TP
\fB\-\-no\-filter\fR
by default, in the interactive modes, httm will filter out files residing upon non\-supported datasets (like ext4, tmpfs, procfs, sysfs, or devtmpfs, etc.), and within any "common" snapshot paths. Here, one may select to disable such filtering. Note, httm will always show the input path, and results from behind any input path when that is the directory path being searched.
.TP
\fB\-\-no\-hidden\fR
do not show information regarding hidden files and directories (those that start with a '.') in the recursive or interactive modes.
.TP
\fB\-\-one\-filesystem\fR
limit recursive search to file and directories on the same filesystem/device as the target directory.
.TP
\fB\-\-no\-traverse\fR
in recursive mode, don't traverse symlinks. Although httm does its best to prevent searching pathologically recursive symlink\-ed paths, here, you may disable symlink traversal completely. NOTE: httm will never traverse symlinks when a requested recursive search is on the root/base directory ("/").
.TP
\fB\-\-no\-live\fR
only display information concerning snapshot versions (display no information regarding live versions of files or directories) in any Display Recursive mode (when DELETED and RECURSIVE are specified, but not an interactive mode). [aliases: \fB\-\-dead\fR, \fB\-\-disco]\fR
.TP
\fB\-\-no\-snap\fR
only display information concerning 'pseudo\-live' versions in any Display Recursive mode (when DELETED and RECURSIVE are specified, but not an interactive mode). Useful for finding the "files that once were" and displaying only those pseudo\-live/zombie files. [aliases: \fB\-\-undead\fR, \fB\-\-zombie]\fR
.TP
\fB\-\-alt\-store=\fR<ALT_STORE>
give priority to specified alternative backups stores, like Restic, and Time Machine.  If this flag is specified, httm will place any discovered alternative backups store as priority snapshots for the root mount point ("/"), ignoring other, potentially more direct, mounts.  Before use, be sure that any such repository is mounted.  You may need superuser privileges to view a repository mounted with superuser permission.  NOTE: httm includes a helper script called "equine" which can assist you in mounting remote and local Time Machine snapshots. [possible values: restic, timemachine]
.TP
\fB\-\-map\-aliases\fR [<MAP_ALIASES>]

src/httm.rs
//       ___           ___           ___           ___
//      /\__\         /\  \         /\  \         /\__\
//     /:/  /         \:\  \        \:\  \       /::|  |
//    /:/__/           \:\  \        \:\  \     /:|:|  |
//   /::\  \ ___       /::\  \       /::\  \   /:/|:|__|__
//  /:/\:\  /\__\     /:/\:\__\     /:/\:\__\ /:/ |::::\__\
//  \/__\:\/:/  /    /:/  \/__/    /:/  \/__/ \/__/~~/:/  /
//       \::/  /    /:/  /        /:/  /            /:/  /
//       /:/  /     \/__/         \/__/            /:/  /
//      /:/  /                                    /:/  /
//      \/__/                                     \/__/
//
// (c) Robert Swinford <robert.swinford<...at...>gmail.com>
//
// For the full copyright and license information, please view the LICENSE file
// that was distributed with this source code.

mod config_helper;
mod deleted;
mod display;
mod interactive;
mod library;
mod lookup;

use crate::config_helper::{install_hot_keys, list_all_filesystems};
use crate::deleted::deleted_exec;
use crate::display::display_exec;
use crate::interactive::interactive_exec;
use crate::library::{for_real_is_dir, read_stdin};
use crate::lookup::lookup_exec;

use clap::{Arg, ArgMatches};
use fxhash::FxHashMap as HashMap;
use rayon::prelude::*;
use std::{
    error::Error,
    fmt,
    fs::{DirEntry, Metadata},
    io::Write,
    path::{Path, PathBuf},
    time::SystemTime,
};

#[derive(Debug)]
pub struct HttmError {
    details: String,
}

impl HttmError {
    fn new(msg: &str) -> HttmError {
        HttmError {
            details: msg.to_owned(),
        }
    }
}

impl fmt::Display for HttmError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.details)
    }
}

impl Error for HttmError {
    fn description(&self) -> &str {
        &self.details
    }
}

#[derive(Clone, Debug, Eq, Hash, PartialEq)]
pub struct PathData {
    system_time: SystemTime,
    size: u64,
    path_buf: PathBuf,
    is_phantom: bool,
}

impl From<&Path> for PathData {
    fn from(path: &Path) -> PathData {
        let metadata_res = std::fs::symlink_metadata(path);
        PathData::new(path, metadata_res)
    }
}

impl From<&DirEntry> for PathData {
    fn from(dir_entry: &DirEntry) -> PathData {
        let metadata_res = dir_entry.metadata();
        let path = dir_entry.path();
        PathData::new(&path, metadata_res)
    }
}

impl PathData {
    fn new(path: &Path, metadata_res: Result<Metadata, std::io::Error>) -> PathData {
        let absolute_path: PathBuf = if path.is_relative() {
            if let Ok(canonical_path) = path.canonicalize() {
                canonical_path
            } else {
                path.to_path_buf()
            }
        } else {
            path.to_path_buf()
        };

        // call symlink_metadata, as we need to resolve symlinks to get non-"phantom" metadata
        let (len, time, phantom) = match metadata_res {
            Ok(md) => {
                let len = md.len();
                let time = md.modified().unwrap_or(SystemTime::UNIX_EPOCH);
                let phantom = false;
                (len, time, phantom)
            }
            // this seems like a perfect place for a None value, as the file has no metadata,
            // however we will want certain iters to print the *request*, say for deleted files,
            // so we set up a dummy Some value just so we can have the path names we entered
            //
            // if we get a spurious example of no metadata in snapshot directories, we just ignore later
            Err(_) => {
                let len = 0u64;
                let time = SystemTime::UNIX_EPOCH;
                let phantom = true;
                (len, time, phantom)
            }
        };

        PathData {
            system_time: time,
            size: len,
            path_buf: absolute_path,
            is_phantom: phantom,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
enum ExecMode {
    Interactive,
    Deleted,
    Display,
}

#[derive(Debug, Clone, PartialEq)]
enum InteractiveMode {
    None,
    Browse,
    Select,
    Restore,
}

#[derive(Debug, Clone, PartialEq)]
enum DeletedMode {
    Disabled,
    Enabled,
    Only,
}

#[derive(Debug, Clone)]
enum SnapPoint {
    Native(Vec<String>),
    UserDefined(UserDefinedDirs),
}

#[derive(Debug, Clone)]
pub struct UserDefinedDirs {
    snap_dir: PathBuf,
    local_dir: PathBuf,
}

#[derive(Debug, Clone)]
pub struct Config {
    paths: Vec<PathData>,
    opt_raw: bool,
    opt_zeros: bool,
    opt_no_pretty: bool,
    opt_no_live_vers: bool,
    opt_recursive: bool,
    exec_mode: ExecMode,
    snap_point: SnapPoint,
    deleted_mode: DeletedMode,
    interactive_mode: InteractiveMode,
    pwd: PathBuf,
    requested_dir: PathData,
}

impl Config {
    fn from(
        matches: ArgMatches,
    ) -> Result<Config, Box<dyn std::error::Error + Send + Sync + 'static>> {
        if matches.is_present("ZSH_HOT_KEYS") {
            install_hot_keys()?
        }
        let opt_zeros = matches.is_present("ZEROS");
        let opt_raw = matches.is_present("RAW");
        let opt_no_pretty = matches.is_present("NOT_SO_PRETTY");
        let opt_no_live_vers = matches.is_present("NO_LIVE");
        let opt_recursive = matches.is_present("RECURSIVE");
        let mut deleted_mode = match matches.value_of("DELETED") {
            None => DeletedMode::Disabled,
            Some("") => DeletedMode::Enabled,
            Some("only") | Some("ONLY") => DeletedMode::Only,
            // invalid value to not specify one of the above
            _ => unreachable!(),
        };
        let mut exec_mode = if matches.is_present("INTERACTIVE")
            || matches.is_present("RESTORE")
            || matches.is_present("SELECT")
        {
            ExecMode::Interactive
        } else if deleted_mode != DeletedMode::Disabled {
            ExecMode::Deleted
        } else {
            ExecMode::Display
        };
        let env_snap_dir = std::env::var_os("HTTM_SNAP_POINT");
        let env_local_dir = std::env::var_os("HTTM_LOCAL_DIR");
        let interactive_mode = if matches.is_present("RESTORE") {
            InteractiveMode::Restore
        } else if matches.is_present("SELECT") {
            InteractiveMode::Select
        } else if matches.is_present("INTERACTIVE") {
            InteractiveMode::Browse
        } else {
            InteractiveMode::None
        };

        if opt_recursive && exec_mode == ExecMode::Display {
            return Err(
                HttmError::new("Recursive search feature only allowed in select modes.").into(),
            );
        }

        // current working directory will be helpful in a number of places
        let pwd = if let Ok(pwd) = std::env::var("PWD") {
            if let Ok(path) = PathBuf::from(&pwd).canonicalize() {
                path
            } else {
                return Err(HttmError::new(
                    "Working directory, as set in your environment, does not appear to exist",
                )
                .into());
            }
        } else {
            return Err(HttmError::new("Working directory is not set in your environment.").into());
        };

        // where is the hidden snapshot directory located?
        // just below we ask whether the user has defined that place
        let raw_snap_var = if let Some(value) = matches.value_of_os("SNAP_POINT") {
            Some(value.to_os_string())
        } else {
            env_snap_dir
        };

        // here we determine how we will obtain our snap point -- has the user defined it
        // or will we find it by searching the native filesystem?
        let snap_point = if let Some(raw_value) = raw_snap_var {
            // user defined dir exists?: check that path contains the hidden snapshot directory
            let path = PathBuf::from(raw_value);
            let hidden_snap_dir = path.join(".zfs").join("snapshot");

            // little sanity check -- make sure the user defined snap dir exist
            let snap_dir = if hidden_snap_dir.metadata().is_ok() {
                path
            } else {
                return Err(HttmError::new(
                    "Manually set mountpoint does not contain a hidden ZFS directory.  Please mount a ZFS directory there or try another mountpoint.",
                ).into());
            };

            // has the user has defined a corresponding local relative directory?
            let raw_local_var = if let Some(raw_value) = matches.value_of_os("LOCAL_DIR") {
                Some(raw_value.to_os_string())
            } else {
                env_local_dir
            };

            // local relative dir can be set at cmdline or as an env var, but defaults to current working directory
            let local_dir = if let Some(value) = raw_local_var {
                let local_dir: PathBuf = PathBuf::from(value);

                // little sanity check -- make sure the user defined local dir exist
                if local_dir.metadata().is_ok() {
                    local_dir
                } else {
                    return Err(HttmError::new(
                        "Manually set local relative directory does not exist.  Please try another.",
                    )
                    .into());
                }
            } else {
                pwd.clone()
            };

            SnapPoint::UserDefined(UserDefinedDirs {
                snap_dir,
                local_dir,
            })
        } else {
            let all_zfs_filesystems = list_all_filesystems()?;

            SnapPoint::Native(all_zfs_filesystems)
        };

        // paths are immediately converted to our PathData struct
        let mut paths: Vec<PathData> =
            if let Some(input_files) = matches.values_of_os("INPUT_FILES") {
                // can unwrap because we confirm "is some" above
                input_files
                    .into_iter()
                    .par_bridge()
                    .map(|string| PathData::from(Path::new(string)))
                    .collect()
            // setting pwd as the path, here, keeps us from waiting on stdin when in non-Display modes
            } else if exec_mode == ExecMode::Interactive || exec_mode == ExecMode::Deleted {
                vec![PathData::from(pwd.as_path())]
            } else if exec_mode == ExecMode::Display {
                read_stdin()?
                    .into_iter()
                    .par_bridge()
                    .map(|string| PathData::from(Path::new(&string)))
                    .collect()
            } else {
                unreachable!()
            };

        // for exec_modes in which we can only take a single directory, process how we handle those here
        let requested_dir: PathData = match exec_mode {
            ExecMode::Interactive => {
                match paths.len() {
                    0 => PathData::from(pwd.as_path()),
                    1 => {
                        match paths.get(0) {
                            Some(pathdata) if for_real_is_dir(&pathdata.path_buf) => {
                                pathdata.to_owned()
                            }
                            Some(pathdata)
                                if pathdata.path_buf.is_file() | pathdata.path_buf.is_symlink() =>
                            {
                                match interactive_mode {
                                    InteractiveMode::Browse | InteractiveMode::None => {
                                        // doesn't make sense to have a non-dir in these modes
                                        return Err(HttmError::new(
                                                "Path specified is not a directory, and therefore not suitable for browsing.",
                                            )
                                            .into());
                                    }
                                    InteractiveMode::Restore | InteractiveMode::Select => {
                                        // non-dir file will just cause us to skip the lookup phase
                                        pathdata.to_owned()
                                    }
                                }
                            }
                            // let's not screw with char, block devices, whatever else.
                            _ => {
                                return Err(HttmError::new(
                                    "Path specified is either not a directory or a file, or does not exist, and therefore is not suitable for an interactive mode.",
                                )
                                .into());
                            }
                        }
                    }
                    n if n > 1 => {
                        return Err(HttmError::new(
                            "May only specify one path in interactive mode.",
                        )
                        .into())
                    }
                    _ => {
                        unreachable!()
                    }
                }
            }
            ExecMode::Deleted => {
                // paths should never be empty for ExecMode::Deleted
                //
                // we only want one dir for a ExecMode::Deleted run, else
                // we should run in ExecMode::Display mode
                match paths.len() {
                    n if n > 1 => {
                        exec_mode = ExecMode::Display;
                        deleted_mode = DeletedMode::Disabled;
                        PathData::from(pwd.as_path())
                    }
                    1 => match paths.get(0) {
                        Some(pathdata) if for_real_is_dir(&pathdata.path_buf) => {
                            pathdata.to_owned()
                        }
                        _ => {
                            exec_mode = ExecMode::Display;
                            deleted_mode = DeletedMode::Disabled;
                            PathData::from(pwd.as_path())
                        }
                    },
                    _ => {
                        // paths should never be empty, but here we make sure
                        PathData::from(pwd.as_path())
                    }
                }
            }
            ExecMode::Display => {
                // in non-interactive mode / display mode, requested dir is just a file
                // like every other file and pwd must be the requested working dir.
                PathData::from(pwd.as_path())
            }
        };

        // deduplicate pathdata and sort if in display mode -- so ./.z* and ./.zshrc only print once
        paths = if exec_mode == ExecMode::Display && paths.len() > 1 {
            let mut unique_paths: HashMap<PathBuf, PathData> = HashMap::default();

            paths.into_iter().for_each(|pathdata| {
                let _ = unique_paths.insert(pathdata.path_buf.clone(), pathdata);
            });

            let mut sorted: Vec<(PathBuf, PathData)> = unique_paths.into_iter().collect();
            sorted.par_sort_unstable_by(|a, b| a.0.cmp(&b.0));

            sorted.into_iter().map(|(_, v)| v).collect()
        } else {
            paths
        };

        let config = Config {
            paths,
            opt_raw,
            opt_zeros,
            opt_no_pretty,
            opt_no_live_vers,
            opt_recursive,
            snap_point,
            exec_mode,
            deleted_mode,
            interactive_mode,
            pwd,
            requested_dir,
        };

        Ok(config)
    }
}

fn parse_args() -> ArgMatches {
    clap::Command::new("httm")
        .about("\nBy default, httm will display non-interactive information about unique file versions contained on ZFS snapshots.\n\n\
        You may also select from the various interactive modes below to browse for, select, and/or restore files.")
        .version("0.7.6") 
        .arg(
            Arg::new("INPUT_FILES")
                .help("in the default, non-interactive mode, put requested files here.  If you enter no files, \
                then httm will pause waiting for input on stdin(3).  In any interactive mode, this is the search path. \
                If none is entered, httm will use the current working directory.")
                .takes_value(true)
                .multiple_values(true)
                .display_order(1)
        )
        .arg(
            Arg::new("INTERACTIVE")
                .short('i')
                .long("interactive")
                .help("interactively browse and search files.")
                .display_order(2)
        )
        .arg(
            Arg::new("SELECT")
                .short('s')
                .long("select")
                .help("interactively browse and search files for selection of a snapshot version.")
                .conflicts_with("RESTORE")
                .display_order(3)
        )
        .arg(
            Arg::new("RESTORE")
                .short('r')
                .long("restore")
                .help("interactively browse and search files for selection of a snapshot version to restore.")
                .conflicts_with("SELECT")
                .display_order(4)
        )
        .arg(
            Arg::new("DELETED")
                .short('d')
                .long("deleted")
                .takes_value(true)
                .default_missing_value("")
                .possible_values(&["only", "ONLY", ""])
                .help("show deleted files in interactive modes, or do a search for all such files, if a directory is specified. \
                If --deleted=only is specified, then, in interactive modes, non-deleted files will be excluded from the search.")
                .display_order(5)
        )
        .arg(
            Arg::new("RECURSIVE")
                .short('R')
                .long("recursive")
                .help("recurse into selected directory to find more files. Only available in interactive and deleted file modes.")
                .display_order(6)
        )
        .arg(
            Arg::new("SNAP_POINT")
                .long("snap-point")
                .help("ordinarily httm will automatically choose your most local snapshot directory, \
                but here you may manually specify your own mount point for that directory, such as the mount point for a remote share.  \
                You can also set via the environment variable HTTM_SNAP_POINT.")
                .takes_value(true)
                .display_order(7)
        )
        .arg(
            Arg::new("LOCAL_DIR")
                .long("local-dir")
                .help("used with SNAP_POINT to determine where the corresponding live root of the ZFS snapshot dataset is.  If not set, \
                httm defaults to your current working directory.  You can also set via the environment variable HTTM_LOCAL_DIR.")
                .requires("SNAP_POINT")
                .takes_value(true)
                .display_order(8)
        )
        .arg(
            Arg::new("RAW")
                .short('n')
                .long("raw")
                .help("display the backup locations only, without extraneous information, delimited by a NEWLINE.")
                .conflicts_with_all(&["ZEROS", "NOT_SO_PRETTY"])
                .display_order(9)
        )
        .arg(
            Arg::new("ZEROS")
                .short('0')
                .long("zero")
                .help("display the backup locations only, without extraneous information, delimited by a NULL CHARACTER.")
                .conflicts_with_all(&["RAW", "NOT_SO_PRETTY"])
                .display_order(10)
        )
        .arg(
            Arg::new("NOT_SO_PRETTY")
                .long("not-so-pretty")
                .help("display the ordinary output, but tab delimited, without any pretty border lines.")
                .conflicts_with_all(&["RAW", "ZEROS"])
                .display_order(11)
        )
        .arg(
            Arg::new("NO_LIVE")
                .long("no-live")
                .help("only display information concerning snapshot versions, and no 'live' versions of files or directories.")
                .display_order(12)
        )
        .arg(
            Arg::new("ZSH_HOT_KEYS")
                .long("install-zsh-hot-keys")
                .help("install zsh hot keys to the users home directory, and then exit")
                .exclusive(true)
                .display_order(12)
        )
        .get_matches()
}

fn main() {
    if let Err(error) = exec() {
        eprintln!("Error: {}", error);
        std::process::exit(1)
    } else {
        std::process::exit(0)
    }
}

fn exec() -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
    let mut out = std::io::stdout();

    // get our program args and generate a config for use
    // everywhere else
    let arg_matches = parse_args();
    let config = Config::from(arg_matches)?;

    let snaps_and_live_set = match config.exec_mode {
        // 1. Do our interactive lookup thing, or not, to obtain raw string paths
        // 2. Determine/lookup whether file matches any files on snapshots
        ExecMode::Interactive => lookup_exec(&config, &interactive_exec(&mut out, &config)?)?,
        ExecMode::Display => lookup_exec(&config, &config.paths)?,
        // deleted_exec is special because it is more convenient to get PathData in 'mod deleted'
        // on raw paths rather than strings, also there is no need to run a lookup on files already on snapshots
        ExecMode::Deleted => deleted_exec(&config, &mut out)?,
    };

    // and display
    let output_buf = display_exec(&config, snaps_and_live_set)?;

    write!(out, "{}", output_buf)?;
    out.flush()?;

    Ok(())
}
